Add separate widgets endpoint

TimothyBJacobs opened the pull request:
## Description
This separates the widgets handling from sidebars into a dedicated widgets controller.

Fixes #25232.

## How has this been tested?
Automated tests in progress.

This will require changes to the client.

## Types of changes
Breaking change.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

TimothyBJacobs committed:
Initial move to widgets endpoint

TimothyBJacobs committed:
Separate sidebar and widgets tests. Add more widgets tests.

TimothyBJacobs commented:
There is some more clean up I'd like to do, but I wanted to get this open so that the changes to the Gutenberg client can be discussed.

As I see it, the Client will now need to do a batch request to update all dirty widgets. Then if the order of widgets has changed, a `PUT` to the sidebar that contains the ordered list of widget IDs like `{ widgets: [ text-1, rss-1 ] }`.

TimothyBJacobs committed:
Fix linting

spacedmonkey commented:
@TimothyBJacobs I can't see this working in postman. 

Don't we need to load the file here https://github.com/WordPress/gutenberg/blob/c1390e0bdd09bd73125317a7f2320ca78726e725/lib/load.php#L35-L37

and setup the class here 
https://github.com/WordPress/gutenberg/blob/c1390e0bdd09bd73125317a7f2320ca78726e725/lib/rest-api.php#L191-L198

If you can do this, I will continue my testing.

TimothyBJacobs commented:
@spacedmonkey yep its completely broken at the moment, sorry for the early ping. Working on a fix.

TimothyBJacobs committed:
Actually register the widgets controller. Register the context param.

spacedmonkey commented on lib/class-wp-rest-widgets-controller.php:
```diff
@@ -0,0 +1,680 @@
+			call_user_func_array( $control['callback'], $arguments );
+			$prepared['rendered_form'] = trim( ob_get_clean() );
+		}
+
```
I would like to see a link to the sidebar.

spacedmonkey commented on lib/class-wp-rest-widgets-controller.php:
```diff
@@ -0,0 +1,684 @@
+				if ( isset( $wp_registered_widgets[ $first_widget_id ] ) ) {
+					$wp_registered_widgets[ $input_widget['id'] ] = $wp_registered_widgets[ $first_widget_id ];
+
+					$widget_class = get_class( $update_control['callback'][0] );
```
Add a unit test for namespaced widget class.

spacedmonkey commented on lib/class-wp-rest-sidebars-controller.php:
```diff
@@ -494,57 +324,7 @@ public function get_item_schema() {
 					'description' => __( 'Nested widgets.', 'gutenberg' ),
```
There seems to be some missing fields. See https://developer.wordpress.org/reference/functions/register_sidebar/#parameters

- class
- before_widget
- after_widget
- before_title
- after_title

TimothyBJacobs committed:
Fix most test failures

spacedmonkey commented on lib/class-wp-rest-widgets-controller.php:
```diff
@@ -0,0 +1,684 @@
+			'id'            => $widget_id,
+			'id_base'       => '',
+			'sidebar'       => $sidebar_id,
+			'widget_class'  => '',
```
Add a unit test for namespaced widget class.

TimothyBJacobs committed:
Fix creating a new widget being given the wrong id.

spacedmonkey commented on lib/class-wp-rest-widgets-controller.php:
```diff
@@ -0,0 +1,695 @@
+					'context'     => array( 'view', 'edit', 'embed' ),
+					'default'     => array(),
+				),
+			),
```
I think it would be useful to add `widget_options`
https://github.com/WordPress/wordpress-develop/blob/cfc3b57488458801c64c750cb500c98c1ef635ad/src/wp-includes/class-wp-widget.php#L166-L172

Specially `customize_selective_refresh`.

TimothyBJacobs replied:
Agreed. We can add this when the widget type endpoint is implemented. No need to add it here now and just remove it in a few days I think.

spacedmonkey commented on lib/class-wp-rest-sidebars-controller.php:
```diff
@@ -116,129 +117,71 @@ public function permissions_check() {
 
+		$request['context'] = 'edit';
+
+		list( , $sidebar ) = $this->get_sidebar( $request['id'] );
```
What is happening on this line?

TimothyBJacobs replied:
The actual sidebar details are returned in the second parameter of the array. I don't really love the `get_sidebar` method, but we can change that after this get's merged since it won't change the output.

spacedmonkey commented on lib/class-wp-rest-sidebars-controller.php:
```diff
@@ -404,33 +249,30 @@ public function prepare_item_for_response( $raw_sidebar, $request ) {
+			$sidebars = wp_get_sidebars_widgets();
+			$widgets  = array_filter(
+				isset( $sidebars[ $sidebar['id'] ] ) ? $sidebars[ $sidebar['id'] ] : array(),
+				function ( $widget_id ) {
+					global $wp_registered_widgets;
+
+					return isset( $wp_registered_widgets[ $widget_id ] );
+				}
```
```suggestion
				function ( $widget_id ) use ( $wp_registered_widgets ) {

					return isset( $wp_registered_widgets[ $widget_id ] );
				}
```

Would this be cleaner?

TimothyBJacobs committed:
Fix linting

spacedmonkey commented on lib/class-wp-rest-widgets-controller.php:
```diff
@@ -0,0 +1,717 @@
+			'sidebar'       => $sidebar_id,
+			'widget_class'  => '',
+			'name'          => $widget['name'],
+			'description'   => ! empty( $widget['description'] ) ? $widget['description'] : '',
```
```suggestion
			'description'   => ! empty( $widget['description'] ) ? $widget['description'] : '',
			'classname'    => ! empty( $widget['classname'] ) ? $widget['classname'] : '',
			'customize_selective_refresh'    => ! empty( $widget['customize_selective_refresh'] ) ? $widget['customize_selective_refresh'] : false,
```

TimothyBJacobs commented:
Huzzah, PHPunit is now passing! Will start on your suggestions @spacedmonkey in a bit.

spacedmonkey committed:
Use widget controller test

spacedmonkey committed:
Add links for widget.

spacedmonkey committed:
Add extra fields.

spacedmonkey committed:
Add links to sidebar.

spacedmonkey committed:
Add extra fields to sidebar.

spacedmonkey committed:
Use use

spacedmonkey commented:
In my testing, a widget with namespace works. 
```
5:
description: "A wibble form for your site."
id: "wibble1-2"
id_base: "wibble1"
name: "Wibble!"
number: 2
rendered: "<section id="wibble1-2" class="widget widget_wibble1"><h2 class="widget-title">TRfsd</h2></section>"
settings: {title: "TRfsd"}
sidebar: "sidebar-1"
widget_class: "WordPressWidgetBoilerplate\WordPress\WP_Widget_Wibble"
_links: {collection: Array(1), self: Array(1), wp:sidebar: Array(1), curies: Array(1)}
__proto__: Object
```

TimothyBJacobs commented:
@spacedmonkey looks like the tests need to be updated for the new fields.

spacedmonkey commented:
@TimothyBJacobs I am thinking about creating that widget_type api and moving some of the fields out into that.

TimothyBJacobs commented:
👍 let's get the tests passing again here first.

adamziel commented:
@TimothyBJacobs this PR removes an endpoint that's currently used in the `edit-widgets` package so in order to merge it we'll have to couple it with, potentially lengthy, JS changes. How about a multi-step approach where we'd just rename the old endpoint to, let's say, `legacy-sidebars`, migrate the React part to the new endpoints, and remove `legacy-sidebars` afterwards?

TimothyBJacobs commented:
If that makes things easier on the JS side of things, that's fine with me! Whatever we can do to make sure it get's done by Beta 1 😀

TimothyBJacobs commented:
@adamziel Another option would be to do what we did for the block directory. We can look for `widgets` in the request of the sidebars endpoint, and do inner requests to the new widgets endpoints. The request/response format for both is the same, so there shouldn't be any issues from that perspective either.

spacedmonkey committed:
Fix tests

spacedmonkey committed:
Add helper method to remove links.

spacedmonkey commented:
@adamziel before you make the changes to JS, I think it is important to think if we want to add a widget type api ( #25969 ). This will change the format of the response. If we are going to add widget type api, we should do it before the js changes, so the changes only have to applied once.

adamziel commented:
> @adamziel Another option would be to do what we did for the block directory. We can look for widgets in the request of the sidebars endpoint, and do inner requests to the new widgets endpoints. The request/response format for both is the same, so there shouldn't be any issues from that perspective either.

@TimothyBJacobs great idea, it sounds even better than the one I proposed 👍 

> @adamziel before you make the changes to JS, I think it is important to think if we want to add a widget type api ( #25969 ). This will change the format of the response. If we are going to add widget type api, we should do it before the js changes, so the changes only have to applied once.

Good call! The inner requests idea proposed by @TimothyBJacobs would make make the entire transition much easier as there would be no multi-step rollout with temporary endpoints.

spacedmonkey commented:
For what is worth, I think this is the format we should go with. 
Sidebar endpoint with embeded widgets, widget with embedded widget type. 
That way you could get in one request, a list of registered sidebars, with all instances of widgets configured in that sidebar with information about the widget type. 

@TimothyBJacobs Do you want to me work on a new PR for this and add yours. I will makes sure to add a unit tests :)

TimothyBJacobs commented:
Yep I agree that format makes sense. I'll add code today to allow the new endpoint to work with the current JS client. Then we can merge this PR.

@spacedmonkey you can get started creating the widget types endpoint in #25969, but don't remove the properties from the widgets/sidebars endpoints yet since that _will_ require the Client code to be updated at the same time. We can then merge the widget types endpoint, update the client JS, and then remove the duplicated fields from the widgets endpoint.

spacedmonkey commented:
@TimothyBJacobs I have a WIP PR for widget types endpoint - https://github.com/WordPress/gutenberg/pull/26042

TimothyBJacobs committed:
Add a "deprecated" mode that uses inner requests to the widgets endpoint.

This allows for sending and receiving full widgets objects in the sidebars
endpoint directly to allow for an easier transition on the client.

TimothyBJacobs committed:
Don't try and render widgets in the inactive sidebar.

Widgets need to be in a valid sidebar to have the necessary context to render.

TimothyBJacobs committed:
Pass "deprecated" parameter to the sidebars endpoint.

TimothyBJacobs commented:
@adamziel I've added support for passing `deprecated=true` as a parameter to the sidebars endpoint. When passed, `widgets` will be expanded to their full entities and saved as full entities.

I've taken a stab at implementing that client side change and it appears to work in my basic testing.

TimothyBJacobs committed:
Fix linting

TimothyBJacobs committed:
The deleted widget cannot be rendered

It is part of the inactive sidebar which doesn't support rendering.

TimothyBJacobs committed:
Move the new endpoints to wp/v2. Make the __experimental sidebars endpoint a proxy to the new endpoint.

TimothyBJacobs committed:
CS

TimothyBJacobs commented:
@adamziel I've changed the approach slightly to move the new widgets and sidebar endpoints to `wp/v2` and now we register explicit shim routes for `__experimental/sidebars` that do inner requests to the proper endpoints.

This means that there aren't any necessary client changes and it should be easier to rip out.

I was having trouble getting `deprecated=true` to be applied in all the necessary places in the data layer.

TimothyBJacobs commented:
The failing test is `test_all_supported` which looks to be failing on `main` as well.

TimothyBJacobs committed:
Don't load non existent sidebars.

noisysocks commented:
I checked out this PR and tested adding, removing and updating blocks and legacy widgets in the widgets editor. The only problem I could spot is that reference widgets (widgets registered without using `WP_Widget`) no longer can be saved.

You can test this by adding this code, going to Appearance → Widgets, inserting a _Marquee Greeting_ into a widget area, then clicking _Save_.

```php
wp_register_sidebar_widget(
	'marquee_greeting',
	'Marquee Greeting',
	function() {
		$greeting = get_option( 'marquee_greeting', 'Hello!' );
		printf( '<marquee>%s</marquee>', esc_html( $greeting ) );
	}
);

wp_register_widget_control(
	'marquee_greeting',
	'Marquee Greeting',
	function() {
		if ( isset( $_POST['marquee-greeting'] ) ) {
			update_option(
				'marquee_greeting',
				sanitize_text_field( $_POST['marquee-greeting'] )
			);
		}

		$greeting = get_option( 'marquee_greeting' );
		?>
		<p>
			<label for="marquee-greeting">Greeting:</label>
			<input
				id="marquee-greeting"
				class="widefat"
				name="marquee-greeting"
				type="text"
				value="<?= esc_attr( $greeting ) ?>"
				placeholder="Hello!"
			/>
		</p>
		<?php
	}
);
```

TimothyBJacobs committed:
Fix adding new reference widgets to a sidebar.

Reference widgets always have an id, so they end up being routed to the
update_item endpoint instead of the create_item endpoint. We now allow for
saving reference widgets that aren't yet assigned to a sidebar if they
include their desired sidebar in the request.

TimothyBJacobs committed:
Merge branch 'add/widgets-endpoint' of github.com:TimothyBJacobs/gutenberg into add/widgets-endpoint

TimothyBJacobs committed:
Make sure widgets are always an array.

TimothyBJacobs commented:
Thanks for testing @noisysocks! I've pushed a fix. The issue was that reference widgets always have an ID, so it was detected as an update. However, the widget wasn't yet assigned to a sidebar which the update endpoint expects. I've fixed this to allow for reference widgets to pass the sidebar id to update and treat that as creation.

noisysocks reviewed (approved):
Tested the widgets editor again and nothing breaks 👍

TimothyBJacobs commented:
Thanks for testing @noisysocks!
